app/connectors/MessageConnector.scala

```diff
@@ -45,58 +47,76 @@ class MessageConnector @Inject() (appConfig: AppConfig, config: Configuration, h
     )
 
     implicit val headerCarrier = hc
       .copy(authorization = None, otherHeaders = Seq.empty)
       .withExtraHeaders(requestHeaders: _*)
 
-    def getHeader(header: String): String =
-      headerCarrier
-        .headersForUrl(headerCarrierConfig)(details.url)
-        .find { case (name, _) => name.toLowerCase == header.toLowerCase }
-        .map { case (_, value) => value }
-        .getOrElse("undefined")
-
-    http
-      .POSTString[HttpResponse](details.url, xml.toString)
-      .map { result =>
-        lazy val logMessage =
-          s"""|Posting NCTS message, ${details.routingMessage}
-              |X-Correlation-Id: ${getHeader("X-Correlation-Id")}
-              |${HMRCHeaderNames.xRequestId}: ${getHeader(HMRCHeaderNames.xRequestId)}
-              |X-Message-Type: ${getHeader("X-Message-Type")}
-              |X-Message-Sender: ${getHeader("X-Message-Sender")}
-              |Accept: ${getHeader("Accept")}
-              |CustomProcessHost: ${getHeader("CustomProcessHost")}
-              |Response status: ${result.status}
+    try {
+      http
+        .POSTString[HttpResponse](details.url, xml.toString)
+        .map { result =>
+          lazy val logMessage =
+            s"""|Posting NCTS message, ${details.routingMessage}
+                |X-Correlation-Id: ${getHeader("X-Correlation-Id", details.url)}
+                |${HMRCHeaderNames.xRequestId}: ${getHeader(HMRCHeaderNames.xRequestId,  details.url)}
+                |X-Message-Type: ${getHeader("X-Message-Type",  details.url)}
+                |X-Message-Sender: ${getHeader("X-Message-Sender",  details.url)}
+                |Accept: ${getHeader("Accept",  details.url)}
+                |CustomProcessHost: ${getHeader("CustomProcessHost",  details.url)}
+                |Response status: ${result.status}
               """.stripMargin
 
-        if (Status.isServerError(result.status) || result.status == Status.FORBIDDEN)
-          logger.warn(logMessage)
-        else
-          logger.info(logMessage)
+          if (Status.isServerError(result.status) || result.status == Status.FORBIDDEN)
+            logger.warn(logMessage)
+          else
+            logger.info(logMessage)
 
-        result
-      }
-      .recover {
-        case e: Exception => {
-          val message = s"${details.url} failed to retrieve data with message ${e.getMessage}"
-          logger.warn(message)
-          HttpResponse(Status.INTERNAL_SERVER_ERROR, message)
+          result
         }
-      }
+    } catch {
+      case e: Exception =>
+        val message = s"${details.url} failed to retrieve data with message ${e.getMessage}"
+        logger.warn(message)
+        Future(HttpResponse(Status.INTERNAL_SERVER_ERROR, message))
+    }
   }
 
+  def postNCTSMonitoring(messageCode: String, timestamp: LocalDateTime,
+                         routingOption: RoutingOption)(implicit hc: HeaderCarrier): Future[HttpResponse] = {
```

Comment:
I think the `headerCarrier` here should just be an explicit parameter as we are working with it directly in the code

Reply:
Yeah this is not a technical objection to be honest, it's more of a style thing - if you work with a piece of data explicitly I think it's better to pass it explicitly, especially since that is what you are doing at the callsite.

Reply:
OK will update this I agree